src/main/java/org/breedinginsight/brapi/v2/BrAPIV2Controller.java

```diff
@@ -79,81 +80,119 @@ public BrAPIServerInfoResponse serverinfo() {
+        AuthenticatedUser actingUser = securityService.getUser();
+
+        logCall(path, request);
+        if (programId != null) {
+            HttpUrl requestUrl = getUrl(programId, path, request);
+
+            var brapiRequest = new Request.Builder().url(requestUrl)
 //                                                    .addHeader("Authorization", "Bearer " + token) //TODO
                                                     .method(method, request.getBody().isPresent() ? RequestBody.create(request.getBody().get(), okhttp3.MediaType.get(MediaType.APPLICATION_JSON)) : null)
                                                     .build();
 
-            try (Response brapiResponse = client.newCall(brapiRequest).execute()) {
-                if(brapiResponse.isSuccessful()) {
-                    try(ResponseBody body = brapiResponse.body()) {
-                        String respBody = body == null ? "" : body.string();
-                        return HttpResponse.ok(respBody);
-                    }
-                } else if(brapiResponse.code() == HttpStatus.NOT_FOUND.getCode()) {
-                    return HttpResponse.notFound();
+            return makeCall(brapiRequest);
+        }
+
+        throw new HttpStatusException(HttpStatus.UNAUTHORIZED, "Unauthorized BrAPI Request");
+    }
+
+    private HttpResponse<String> makeCall(Request brapiRequest) {
+        OkHttpClient client = new OkHttpClient();
+        try (Response brapiResponse = client.newCall(brapiRequest).execute()) {
+            if(brapiResponse.isSuccessful()) {
+                try(ResponseBody body = brapiResponse.body()) {
+                    String respBody = body == null ? "" : body.string();
+                    return HttpResponse.ok(respBody);
+                } catch (Exception e) {
+                    return HttpResponse.ok("");
                 }
-            } catch (IOException e) {
-                log.error("Error calling BrAPI Service", e);
-                throw new HttpStatusException(HttpStatus.INTERNAL_SERVER_ERROR, "Error calling BrAPI Service");
+            } else {
+                return HttpResponse.status(HttpStatus.valueOf(brapiResponse.code()));
             }
+        } catch (IOException e) {
+            log.error("Error calling BrAPI Service", e);
+            throw new HttpStatusException(HttpStatus.INTERNAL_SERVER_ERROR, "Error calling BrAPI Service");
         }
+    }
 
-        throw new HttpStatusException(HttpStatus.UNAUTHORIZED, "Unauthorized BrAPI Request");
+    private HttpUrl getUrl(UUID programId, String path, HttpRequest<?> request) throws NotFoundException {
+
+        ProgramBrAPIEndpoints programBrAPIEndpoints;
+        try {
+            programBrAPIEndpoints = programService.getBrapiEndpoints(programId);
+        } catch (DoesNotExistException e) {
+            throw new NotFoundException("Program does not exist");
```

Comment:
Should this be an HttpStatusException with HttpStatus.NOT_FOUND?

Reply:
Yeah probably, I'll add that.

Reply:
changed

Reply:
can you get rid of the throws NotFoundException on the method signature?

Reply:
removed